File: multithreaded_server_provisioner.go
package main

import (
	"fmt"
	"log"
	"os"
	"strings"
	"sync"
	"time"

	"golang.org/x/crypto/ssh"
)

// SSH configuration
const (
	sshHost     = "your_instance_ip"
	sshPort     = "22"
	sshUser     = "your_ssh_username"
	sshPassword = "your_ssh_password"
)

// Define a function to execute SSH commands
func sshExec(host, username, password, command string) {
	client, err := ssh.Dial("tcp", host+":"+sshPort, &ssh.ClientConfig{
		User: username,
		Auth: []ssh.AuthMethod{ssh.Password(password)},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
	})
	if err != nil {
		log.Fatalf("Failed to dial: %v", err)
	}
	defer client.Close()

	session, err := client.NewSession()
	if err != nil {
		log.Fatalf("Failed to create session: %v", err)
	}
	defer session.Close()

	// Execute the command
	output, err := session.CombinedOutput(command)
	if err != nil {
		log.Fatalf("Command failed: %v", err)
	}
	fmt.Printf("[%s] %s", host, strings.TrimSpace(string(output)))
}

func main() {
	var wg sync.WaitGroup

	// List of tasks to be executed on the instance
	tasks := []string{
		"sudo apt-get install -y vim curl git unzip",
		"sudo ufw allow 22/tcp",
		"sudo apt-get install -y mysql-server",
		"git clone https://github.com/rbenv/rbenv.git ~/.rbenv",
		`echo 'export PATH="$HOME/.rbenv/bin:$PATH"' >> ~/.bashrc`,
		`echo 'eval "$(rbenv init -)"' >> ~/.bashrc`,
	}

	// Update task (execute before the goroutines)
	cmd := "sudo apt-get update"
	cmdOutput, err := exec.Command("bash", "-c", cmd).CombinedOutput()
	if err != nil {
		log.Fatalf("Command failed: %v", err)
	}
	fmt.Printf("%s", cmdOutput)

	// Create a Goroutine for each task
	for _, task := range tasks {
		wg.Add(1)
		go func(task string) {
			defer wg.Done()
			sshExec(sshHost, sshUser, sshPassword, task)
		}(task)
	}

	// Wait for all Goroutines to finish
	wg.Wait()

	fmt.Println("Instance provisioning completed.")
}

File: multithreaded_server_provisioner.rb
#!/usr/bin/env ruby

require 'open3'
require 'net/ssh'
require 'thread'

# Define a function to execute system commands
def run_command(command)
  stdout, stderr, status = Open3.capture3(command)
  puts stdout
  raise "Command failed: #{stderr}" unless status.success?
end

# Define a function to SSH into the instance and execute commands
def ssh_exec(host, username, password, commands)
  Net::SSH.start(host, username, password: password) do |ssh|
    commands.each do |cmd|
      ssh.exec!(cmd) do |channel, stream, data|
        puts "[#{host}] #{data}"
      end
    end
  end
end

# Configuration
instance_ip = 'your_instance_ip'
ssh_username = 'your_ssh_username'
ssh_password = 'your_ssh_password'

# List of tasks to be executed on the instance
tasks = [
  "sudo apt-get update",
  "sudo apt-get install -y vim curl git unzip",
  "sudo ufw allow 22/tcp",
  "sudo apt-get install -y mysql-server",
  "git clone https://github.com/rbenv/rbenv.git ~/.rbenv",
  "echo 'export PATH=\"$HOME/.rbenv/bin:$PATH\"' >> ~/.bashrc",
  "echo 'eval \"$(rbenv init -)\"' >> ~/.bashrc"
]

# Create an array of threads to execute tasks concurrently
threads = tasks.map do |task|
  Thread.new do
    ssh_exec(instance_ip, ssh_username, ssh_password, [task])
  end
end

# Start the threads and wait for them to finish
threads.each(&:join)

puts "Instance provisioning completed."
